### ISSUE TITLE
Target tab completion invokes `rustup`

### ISSUE THREAD

**Issue by mkoncek** [rust-lang/cargo#12585]
### Problem

Invoking `cargo build --target <TAB>` invokes `rustup`. Some environments may not have it installed and I do not see why it should be invoked if the tab completion can run equivalent of `rustc --print target-list`.

### Steps

1. Set up a basic Rust project with cargo
2. Write `cargo build --target ` on terminal and then press the TAB key.
3. If `rustup` is not installed, terminal will print:
`cargo build --target bash: rustup: command not found`

### Possible Solution(s)

I suppose `cargo` could return the output of `rustc --print target-list`

### Notes

It is generally a good idea to depend only on the most basic tools, especially if they get the job done right, like `rustc` does.

### Version

```text
cargo 1.71.0
release: 1.71.0
host: x86_64-unknown-linux-gnu
libgit2: 1.6.4 (sys:0.17.1 system)
libcurl: 8.0.1 (sys:0.4.61+curl-8.0.1 system ssl:OpenSSL/3.0.9)
os: Fedora 38.0.0 [64-bit]
```
```

**Comment by epage** [comment]
Interesting, the zsh completions are using `rustc --print target-list` but the bash ones aren't.

zsh support for targets was added in #8740

bash support was added in #6640

I don't see any references to why different strategies were taken, so it seems like we could consolidate down to just `rustc`

**Comment by weihanglo** [comment]
Another idea is we can go with `rustup` way first and fall back to `rustc --print target-list`.

**Comment by epage** [comment]
Is there a reason to prefer rustup?  I guess so you can check which targets are actually installed?

**Comment by weihanglo** [comment]
Yes. Completing with installed ones. `rustc --print target-list` now prints 218 target triples which doesn't sounds too friendly.

**Comment by Angelin01** [comment]
I can probably whip this up by tomorrow, already have the bash one ready, just need to do some testing with the ZSH one.
